### PR TITLE
docs(*): Change doc generation to keep them out of master

### PR DESCRIPTION
**Pull request by ProbablePrime**
<details> 
  <summary>Ok, so docs being in master were really annoying and they often became out of date, this PR shunts them into the gh-pages branch and will automaticaly update them as a part of a release cycle
</summary>
    oK, so dOcS beInG In mAStEr wErE REaLLy aNnoyINg and tHEy OftEn becAmE OUt Of daTe, thIS Pr ShUNTS theM iNTo ThE gh-PageS BRANch and Will auTOmATICAlY UpDAte them aS A parT of a ReLEAse cYCLe
</details>
<br />

Oh and typescript 2.4 is kinda mean: http://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html

### CODE REVIEW

File: src/state/interfaces/controls/IJoystick.ts [hunk 1]
```diff
@@ -75,6 +75,8 @@ export interface IJoystick extends IControl, IJoystickData {
      */
     update(controlUpdate: IJoystickUpdate): Promise<void>;
 
+    giveInput(input: IJoystickInput): Promise<void>;
```
Comment (UnwrittenFun) [inline src/state/interfaces/controls/IJoystick.ts]: .

File: package.json [hunk 2]
```diff
@@ -24,9 +24,10 @@
     "build": "npm run build:ts && npm run build:webpack && npm run build:examples",
     "test": "npm run lint && npm run test:unit",
     "test:unit": "mocha --compilers ts:ts-node/register \"src/**/*.spec.ts\"",
-    "prepublish": "npm run build",
+    "prepublishOnly": "npm run build && npm run docs:publish",
```
Comment (SimonSchick) [inline package.json]: Keep in mind this only works with npm5.

File: src/state/controls/Button.ts [hunk 3]
```diff
@@ -94,7 +94,8 @@ export class Button extends Control<IButtonData> implements IButton {
      */
     public update(controlUpdate: IButtonUpdate): Promise<void> {
         // Clone to prevent mutations
-        const changedData = { ...controlUpdate };
+        // XXX: Typescript 2.4 is strict, let the compiler be clever.
+        const changedData = Object.assign({}, controlUpdate );
```
Comment (SimonSchick) [inline src/state/controls/Button.ts]: Hu? This should work.

Comment (UnwrittenFun) [inline src/state/controls/Button.ts]: it works but requires you to typecast `changedData` to `IButtonUpdate`

Comment (ProbablePrime) [inline src/state/controls/Button.ts]: It does but it requires a cast, i prefer it without the cast.

Comment (SimonSchick) [inline src/state/controls/Button.ts]: Compiles to the same in the end so I guess.

**Pull request merged**